### PR TITLE
Custom 404 error handler

### PR DESCRIPTION
**Pull request by fabricedesre**

### CODE REVIEW

**Comment by fabricedesre** [comment]
Rebasing an old patch, this gives us a more consistent 404 error.
r? @ferjm

File: src/http_server.rs [hunk 1]
```diff
@@ -3,13 +3,38 @@
...
-use static_router;
 use foxbox_users::users_router::UsersRouter;
-use iron::Iron;
+use iron::{ Iron, Request, Response, IronResult, AfterMiddleware, Chain };
+use iron::error::{ IronError };
+use iron::status::Status;
 use mount::Mount;
+use router::NoRoute;
+use service_router;
+use static_router;
```
Comment (ferjm) [inline src/http_server.rs:13]: Thanks :)

File: src/http_server.rs [hunk 2]
```diff
@@ -3,13 +3,38 @@
...
+struct Custom404;
+
+impl AfterMiddleware for Custom404 {
+    fn catch(&self, _: &mut Request, err: IronError) -> IronResult<Response> {
+        use std::io::Error as StdError;
+        use std::io::ErrorKind;
+
+        if let Some(_) = err.error.downcast::<NoRoute>() {
+            // Router error
+            return Ok(Response::with((Status::NotFound, format!("Unknown resource: {}", err))));
```
Comment (ferjm) [inline src/http_server.rs]: nit: split in two lines, please

File: src/http_server.rs [hunk 3]
```diff
@@ -3,13 +3,38 @@
...
+        } else if let Some(err) = err.error.downcast::<StdError>() {
+            // StaticFile error
+            if err.kind() == ErrorKind::NotFound {
+                return Ok(Response::with((Status::NotFound, format!("Unknown resource: {}", err))));
```
Comment (ferjm) [inline src/http_server.rs]: ditto

**Comment by ferjm** [comment]
LGTM. Write a test or file a bug for them, please. Thank you!

**Comment by fabricedesre** [comment]
I filed https://github.com/fxbox/foxbox/issues/175 to add tests.

**Pull request merged**